lib/Conversion/TorchToTosa/TorchToTosa.cpp

```diff
@@ -1132,6 +1133,34 @@ Type getMatMulOutputType(Type inputElemTy, PatternRewriter &rewriter) {
   return outputElemTy;
 }
 
+RankedTensorType getCastedInputTypeForMatmul(Value inputValue,
+                                             PatternRewriter &rewriter) {
+  // Check to see if the inputs to the matmul where casted from another type
+  auto preCastType =
+      TypeSwitch<Operation *, RankedTensorType>(inputValue.getDefiningOp())
+          .Case([](AtenToDtypeOp op) {
+            return cast<RankedTensorType>(op->getOperand(0).getType());
+          })
+          .Case([](tosa::CastOp op) {
+            return cast<RankedTensorType>(op->getOperand(0).getType());
+          })
+          .Default([](Operation * /*op*/) { return RankedTensorType(); });
+  if (!preCastType) {
+    return preCastType;
+  }
+  // Calculate the expected accumulator type based on the input type of the cast
+  auto accumulatorType =
+      getMatMulOutputType(preCastType.getElementType(), rewriter);
+  // If the expected accumulatorType for the given input type to the cast
+  // matches the output type of the cast then we can fold the casting into the
+  // matmul. Because the casting is an up-cast and does not affect the numeric
+  // values due to rounding or saturation.
```

Comment:
```suggestion
  // matmul.
```
You can actually get different results when calculating in a bigger data type, because you are "too precise", i.e., errors propagate differently. However, this transformation here is still fine, since the tosa matmul operator defines that as a first step, the inputs are converted to the output type. So you just go from the explicit casts to relying on the tosa operator semantics to do the cast.
Maybe something like:
`The tosa matmul is defined to cast the inputs to the output type first, so we do not need explicit casts up front.`

Reply:
comment updated.